### PR TITLE
Ticket/15915

### PR DESCRIPTION
**Pull request by robertwb**
This pull request is just for comments.

### CODE REVIEW

File: src/sage/crypto/lwe.py [hunk 1]
```diff
@@ -946,19 +774,16 @@ def balance_sample(s, q=None):
 
         sage: from sage.crypto.lwe import balance_sample, samples, Regev
         sage: map(balance_sample, samples(10, 5, Regev))
-        [((-9, -4, -4, 4, -4), 6), ((-3, -10, 8, -3, -1), -10), ((-6, -12, -3, -2, -6), -6),
-        ...
-        ((-1, -8, -11, 13, 4), -6), ((10, 11, -3, -13, 0), 6), ((6, -1, 2, -11, 14), 2)]
+        [((-9, -4, -4, 4, -4), 4), ((-8, 11, 12, -11, -11), -6), ((-13, -7, -12, 7, 7), -10), ((-6, -12, -3, -2, -6), -7), ((-5, -10, 1, 5, -5), -10), ((-14, 5, 12, 14, -9), 1), ((12, 11, 13, -1, 9), 10), ((1, -6, 2, -10, 4), 7), ((-11, 12, 0, -6, -3), 7), ((-7, 14, 8, 11, -8), -11)]
```
Comment (robertwb) [inline src/sage/crypto/lwe.py:777]: I prefer the line break with ... here and below.

File: src/sage/rings/integer_ring.pyx [hunk 2]
```diff
@@ -682,6 +683,14 @@ cdef class IntegerRing_class(PrincipalIdealDomain):
           If the distribution ``'mpz_rrandomb'`` is specified, the output is
           in the range from 0 to `2^x - 1`.
 
+          If the distribution ``'gaussian'`` is specified, the output is sampled
+          from a discrete Gaussian distribution with parameter sigma=x and centered
+          at zero. That is, the integer v is returned with probability proportional
+          to exp(-v^2/(2*sigma^2)). See :mod:`sage.stats.distributions.discrete_gaussian_integer` for details.
+          Note that if many samples from the same discrete Gaussian distribution
+          are needed, it is much faster to construct a DiscreteGaussianSampler object
+          which is then repeatedly queried.
```
Comment (robertwb) [inline src/sage/rings/integer_ring.pyx:692]: Idea: could we cache the last created distribution so calling it with the exact same params is fast?

File: src/sage/stats/distributions/__init__.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+import discrete_gaussian_integer
+import discrete_gaussian_polynomial
```
Comment (robertwb) [inline src/sage/stats/distributions/__init__.py:2]: missing newline

File: src/sage/stats/distributions/discrete_gaussian_integer.pyx [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+          integer-valued `c` are supported.
+
+        - ``"uniform+online"`` - samples are drawn from a uniform distribution and
+          accepted with probability proportional to `\exp(-(x-c)^2/(2σ^2))` where
+          `\exp(-(x-c)^2/(2σ^2))` is computed in each invocation. Typically this
+          is very slow.  See [DDLL13]_ for details.  Any real-valued `c` is
+          accepted.
+
+        - ``"sigma2+logtable"`` - samples are drawn from an easily samplable
+          distribution `k·σ_2` and accepted with probability proportional to
```
Comment (robertwb) [inline src/sage/stats/distributions/discrete_gaussian_integer.pyx:104]: What is \sigma_2?

File: src/sage/stats/distributions/discrete_gaussian_integer.pyx [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+            (True, True, 2.45...)  
+
+            sage: from sage.stats.distributions.discrete_gaussian_integer import DiscreteGaussianIntegerSampler
+            sage: D = DiscreteGaussianIntegerSampler(1.0, c=2.5, tau=6, precision="dp")
+            sage: l = [D() for _ in xrange(2^18)]
+            sage: min(l), max(l), abs(mean(l)-2.5) < 0.01
+            (-2, 7, True)
+
+        These generators cache random bits for performance reasons. Hence, resetting
+        the seed of the PRNG might not have the expected outcome::
```
Comment (robertwb) [inline src/sage/stats/distributions/discrete_gaussian_integer.pyx:191]: Can one call reset (or similar) to flush this cache?

File: src/sage/stats/distributions/discrete_gaussian_integer.pyx [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+            self._gen_mp = NULL
+            self._sigma = RR(sigma)
+            self._c = RR(c)
+        else:
+            raise ValueError("Parameter precision '%s' not supported."%precision)
+            
+        self._tau = Integer(tau)
+        self._algorithm = algorithm_str
+
+    def __clear__(self):
```
Comment (robertwb) [inline src/sage/stats/distributions/discrete_gaussian_integer.pyx:271]: Did you mean **del**?

File: src/sage/stats/distributions/discrete_gaussian_integer.pyx [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+        """
+        Gaussian parameter sigma.
+        
+        EXAMPLE::
+
+            sage: from sage.stats.distributions.discrete_gaussian_integer import DiscreteGaussianIntegerSampler
+            sage: D = DiscreteGaussianIntegerSampler(3.0, 2); D.sigma
+            3.0...
+        """
+        return self._sigma
```
Comment (robertwb) [inline src/sage/stats/distributions/discrete_gaussian_integer.pyx:335]: Why not make these cdef readonly attributes, without the leading underscore?

File: src/sage/rings/integer_ring.pyx (hunk 2 above)
Comment (malb) [inline src/sage/rings/integer_ring.pyx:692]: done

File: src/sage/stats/distributions/__init__.py (hunk 3 above)
Comment (malb) [inline src/sage/stats/distributions/__init__.py:2]: fixed

File: src/sage/stats/distributions/discrete_gaussian_integer.pyx (hunk 4 above)
Comment (malb) [inline src/sage/stats/distributions/discrete_gaussian_integer.pyx:104]: I added more documentation

File: src/sage/stats/distributions/discrete_gaussian_integer.pyx (hunk 5 above)
Comment (malb) [inline src/sage/stats/distributions/discrete_gaussian_integer.pyx:191]: I added this.

File: src/sage/stats/distributions/discrete_gaussian_integer.pyx (hunk 6 above)
Comment (malb) [inline src/sage/stats/distributions/discrete_gaussian_integer.pyx:271]: I meant **dealloc**

File: src/sage/stats/distributions/discrete_gaussian_integer.pyx (hunk 7 above)
Comment (malb) [inline src/sage/stats/distributions/discrete_gaussian_integer.pyx:335]: done

**Pull request merged**